go/vt/topo/test/shard.go

```diff
@@ -94,3 +96,72 @@ func checkShard(t *testing.T, ts *topo.Server) {
 		t.Errorf("GetShardNames(666): %v", err)
 	}
 }
+
+// checkShardWithLock verifies the Shard operations work correctly With Lock
+func checkShardWithLock(t *testing.T, ts *topo.Server) {
+	ctx := context.Background()
+	if err := ts.CreateKeyspace(ctx, "test_keyspace", &topodatapb.Keyspace{}); err != nil {
+		t.Fatalf("CreateKeyspace: %v", err)
+	}
+
+	unblock := make(chan struct{})
+	finished := make(chan struct{})
+
+	// Check GetShardNames returns [], nil for existing keyspace with no shards.
+	if names, err := ts.GetShardNames(ctx, "test_keyspace"); err != nil || len(names) != 0 {
+		t.Errorf("GetShardNames(keyspace with no shards) didn't return [] nil: %v %v", names, err)
+	}
+
+	if err := ts.CreateShard(ctx, "test_keyspace", "b0-c0"); err != nil {
+		t.Fatalf("CreateShard: %v", err)
+	}
+
+	_, unlock1, err := ts.LockShard(ctx, "test_keyspace", "b0-c0", "lock")
+	if err != nil {
+		t.Errorf("CreateShard called second time, got: %v", err)
+	}
+
+	// As soon as we're unblocked, we try to lock the keyspace.
+	go func() {
+		<-unblock
+		waitUntil := time.Now().Add(10 * time.Second)
+		var isUnLocked1 = false
+		for time.Now().Before(waitUntil) {
+			_, unlock2, err := ts.TryLockShard(ctx, "test_keyspace", "b0-c0", "lock")
+			if err != nil {
+				if !topo.IsErrType(err, topo.NodeExists) {
+					require.Fail(t, "expected node exists during trylock", err.Error())
+				}
+				time.Sleep(1 * time.Second)
+				var finalErr error
+				unlock1(&finalErr)
+				isUnLocked1 = true
+				if finalErr != nil {
+					require.Fail(t, "Unlock(test_keyspace) failed", finalErr.Error())
+				}
+			} else {
```

Comment:
Shouldn't this be considered a test failure as the Try should have failed, right? That's what we're testing.

Reply:
thats what we are trying to do here. We are doing TryLockShard and expect NodeExists error until it prior lock gets unlock and then we expect to acquire lock in else case.

Reply:
Unless I'm misreading things, the test would currently pass if:
1.  TryLock fails with a NodeExists error
2. TryLock succeeds 

And it only fails if TryLock fails with a different error. So what are we trying to test and confirm here? I thought the test should only pass if 1 is true, no?

Reply:
We walked through this together and it now makes sense after the refactoring. We'll be adding comments to make it even more obvious to the reader.